### PR TITLE
Feature/topic filter generator agent singleton

### PR DESCRIPTION
**Pull request by sauroter**

### CODE REVIEW

File: src/main/java/org/varroa/chunk/ChunkGeneratorUtil.java [hunk 1]
```diff
@@ -49,39 +50,26 @@
...
-
-        // calculate the ratio of topics to clients
-        float topicsPerClient = ((float) topicCount) / ((float) clientsInClientGroup);
-
-        // the minimum is one topic filter per client
-        if (topicsPerClient < 1.0F) {
-            topicsPerClient = 1.0F;
+    @NotNull
+    static TopicFilterGeneration getTopicFilterGeneration(
+            final int clientsInClientGroup, final @NotNull SubscriptionEntity subscriptionEntity) {
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/chunk/ChunkGeneratorUtil.java]: clientsInClientGroup -> clientCountInGroup to be consistent

File: src/main/java/org/varroa/chunk/LifeCycleProcessor.java [hunk 2]
```diff
@@ -188,41 +169,18 @@
...
-     * @param clientCountInChunk     amount of clients in this chunk
-     * @param clientOffset           the current chunk offset
      * @return the finished SubscribeCommand
      */
     private @NotNull SubscribeCommand processSubscribe(
             final @NotNull SubscribeCommandEntity subscribeCommandEntity,
-            final int clientCountInGroup,
-            final int clientCountInChunk,
-            final int clientOffset) {
+            final int clientCountInGroup) {
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/chunk/LifeCycleProcessor.java]: clientCountInGroup is not needed here

Comment (RobinAtherton) [reply src/main/java/org/varroa/chunk/LifeCycleProcessor.java]: 🔥 🔥 🔥

File: src/main/java/org/varroa/chunk/command/generation/TopicFilterGenerators.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                .expireAfterAccess(10, TimeUnit.MINUTES)
+                .build();
+    }
+
+    public @NotNull TopicFilterGenerator get(
+            final @NotNull String subscriptionId,
+            final @NotNull TopicFilterGeneration generation) {
+
+
+        final TopicFilterGenerator optimistic = generators.getIfPresent(subscriptionId);
```
Comment (RobinAtherton) [inline src/main/java/org/varroa/chunk/command/generation/TopicFilterGenerators.java]:
```suggestion
        final TopicFilterGenerator optimisticGenerator = generators.getIfPresent(subscriptionId);
```

File: src/main/java/org/varroa/chunk/command/generation/TopicFilterGenerator.java [hunk 4]
```diff
@@ -12,8 +12,8 @@
     /**
      * Generates a List of topicFilters for each MqttAgent.
      *
-     * @param positionInChunk the relative offset of the MqttAgent in the Chunk
+     * @param clientOffset the relative offset of the MqttAgent in the Client Group
      * @return a non empty list of topic filters
      */
-    @NotNull List<String> generate(final int positionInChunk);
+    @NotNull List<String> generate(final int clientOffset, final int clientsInClientGroup);
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/chunk/command/generation/TopicFilterGenerator.java]: clientsInClientGroup -> clientCountInGroup

File: src/main/java/org/varroa/chunk/command/generation/TopicFilterGenerators.java (hunk 3 above)
Comment (sauroter) [reply src/main/java/org/varroa/chunk/command/generation/TopicFilterGenerators.java]: nope

**Pull request merged**